feat(sdp): Updating the ip to endpoint table for sdp

vipul-21 opened the pull request:
- Adds the ip to endpoint info i.e identity and endpoint id in the table.
- Adds a watcher that listens for changes in the table and updates the prefixs for faster lookup.
- Lookup handlers for DNS proxy for serving the DNS req.
- Send the identity<>cidr map from the cilium agent. These cidr are not associated with
an endpoint.

Fixes: https://github.com/cilium/cilium/issues/30984
CFP: https://github.com/cilium/design-cfps/pull/54
Relevant PRs: #40982, #41507 
```release-note
feat(sdp): Updating the ip to endpoint table for sdp
```

vipul-21 committed:
feat(sdp): Updating the ip to endpoint table for sdp

- Adds the ip to endpoint info i.e identity and endpoint id in the table.
- Adds a watcher that listens for changes in the table and updates the prefixs for
faster lookup.
- Lookup handlers for DNS proxy for serving the DNS req.
- Send the identity<>cidr map from the cilium agent. These cidr are not associated with
an endpoint.

Signed-off-by: Vipul Singh <singhvipul@microsoft.com>

vipul-21 commented:
/test

pchaigno commented:
> no open GH issue. Found a closed one: #34356. The issue should have been fixed with the PR: #34388 though. cc: @pchaigno I don't have the permissions to reopen the issue.(Can open a new one if needed)

Let's open a new one. We confirmed that PR fixed the issue, so it resurfaced in another way. Maybe https://github.com/cilium/cilium/commit/c49d46ac3b29d7cf3a0fd07a79d502183a39a086 broke it?

vipul-21 commented:
@squeed Friendly ping for the review on the PR. Thanks !
cc: @cilium/fqdn

squeed commented on pkg/fqdn/service/service.go:
```diff
@@ -306,6 +312,12 @@ func (s *FQDNDataServer) sendAndRecvAckForDNSPolicies(stream pb.FQDNData_StreamP
 		}
+		if len(prefixWithoutEndpoint) > 0 {
+			// Add IPs without endpoint info i.e. cidrs that do not map to any endpoint
+			endpointInfos = append(endpointInfos, &pb.EndpointInfo{
```
This isn't an endpoint. Why does it have IPs?

vipul-21 replied:
Addressed this by adding a separate table for identity mappings.

vipul-21 commented on standalone-dns-proxy/pkg/client/client.go:
```diff
@@ -110,11 +126,53 @@ func newDNSRulesTable(db *statedb.DB) (statedb.RWTable[service.PolicyRules], err
 }
+
+// updatePolicyState processes the received PolicyState message and updates the DNSRules/IPToEndpoint table accordingly.
+func (c *GRPCClient) updatePolicyState(state *pb.PolicyState) error {
```
It is currently called on every update on the endpoint. 
Agent Side : The rules are updated in the table for every endpoint and we listen on the tables for the updates and send the snapshot of the policy state to sdp.
We should ideally send the deltas but to avoid the different operation(upsert/delete) we kept the snapshot based approach plus the rate limiter.  We did discuss to take this as an improvement once we have the core feature added.
https://github.com/cilium/cilium/pull/39220#discussion_r2068497964

vipul-21 committed:
Use separate tables for prefixToIdentity and IPToEndpoint mappings in SDP

- Introduce separate tables for prefix-to-identity mapping and IP-to-endpoint mapping.
- Update the SDP lookup function to use these new tables when resolving endpoints and
identities during DNS request processing.
- Modify the existing protobuf definition for the standalone DNS proxy to send the
prefix-to-identity mapping separately.

Signed-off-by: Vipul Singh <singhvipul@microsoft.com>

vipul-21 commented:
/test

squeed commented on pkg/fqdn/service/service.go:
```diff
@@ -274,14 +274,14 @@ func (s *FQDNDataServer) sendAndRecvAckForDNSPolicies(stream pb.FQDNData_StreamP
 		var prefixes []netip.Prefix
 		endpointToIPsBytes := make(map[uint64][][]byte) // Group IPs by endpoint ID for this identity
-
+		prefixWithoutEndpoint := [][]byte{}
```
not sure we need this? we should send all prefix -> ID mappings, even if that prefix happens to correspond to a pod on the node.

squeed replied:
I still see prefixWithoutEndpoint?

vipul-21 committed:
send all prefix to id mappings to sdp

Signed-off-by: Vipul Singh <singhvipul@microsoft.com>

vipul-21 commented:
/test

vipul-21 commented on standalone-dns-proxy/pkg/lookup/lookup.go:
```diff
@@ -4,21 +4,31 @@
+	db                    *statedb.DB
+	// prefixLengths tracks the unique set of prefix lengths for IPv4 and
+	// IPv6 addresses in order to optimize longest prefix match lookups.
+	prefixLengths *counter.PrefixLengthCounter
```
@squeed Can we go ahead with tracking of `prefixLengths` for now and once we have the `StateDB` implementation we can update it. ?

joamaki replied:
I would suggest we'd do that since it'll take time for the StateDB changes to happen.

vipul-21 commented on standalone-dns-proxy/pkg/lookup/lookup.go:
```diff
@@ -27,12 +37,97 @@ func (r *rulesClient) LookupByIdentity(nid identity.NumericIdentity) []string {
+	if _, cidr, err := net.ParseCIDR(prefix.String()); err == nil {
+		ones, bits := cidr.Mask.Size()
+		if ones == bits {
+			info, _, exists := r.prefixToIdentityTable.Get(r.db.ReadTxn(), client.PrefixToIdentityIndex.Query(prefix))
```
Updated it.

vipul-21 commented:
/test

vipul-21 commented on standalone-dns-proxy/pkg/lookup/lookup_test.go:
```diff
@@ -0,0 +1,115 @@
+
+	// Pre-insert an entry for testing
+	insertIP(db, table, netip.MustParsePrefix("10.0.0.1/32"), 123, identity.NumericIdentity(5))
+	insertIP(db, table, netip.MustParsePrefix("10.0.0.0/24"), 42, identity.NumericIdentity(5))
```
I updated the test cases.

vipul-21 commented:
/test

vipul-21 commented:
/ci-gateway-api

vipul-21 commented:
[Cilium E2E Upgrade (ci-e2e-upgrade)](https://github.com/cilium/cilium/actions/runs/19086048868) - failing due to a warning log that has a retry mechanism:
>time=2025-11-04T23:38:20.750321606Z level=warn source=/go/src/github.com/cilium/cilium/pkg/datapath/orchestrator/orchestrator.go:229 msg="Failed to initialize datapath, retrying later" module=agent.datapath.orchestrator error="inserting ipv4 proxy route {proxy 2004 0.0.0.0/0 0 100 false invalid IP 0.0.0.0 0xc0026b80a0 0 universe local {{0 0 <nil>} <nil> 0 {0}}}: timeout waiting for parameter proxy:2004:0.0.0.0/0 reconciliation: %!w(<nil>)" retryDelay=10s (1 occurrences)

vipul-21 commented:
/ci-e2e-upgrade
